checker/src/org/checkerframework/checker/optional/OptionalVisitor.java

```diff
@@ -0,0 +1,340 @@
+    /** @return true iff expression is a call to java.util.Optional.get */
+    private boolean isCallToGet(ExpressionTree expression) {
+        return OptionalUtils.isMethodInvocation(expression, "get", 0, atypeFactory);
+    }
+
+    /** @return true iff expression is a call to java.util.Optional.isPresent */
+    private boolean isCallToIsPresent(ExpressionTree expression) {
+        return OptionalUtils.isMethodInvocation(expression, "isPresent", 0, atypeFactory);
+    }
+
+    /** @return true iff expression is a call to java.util.Optional.of */
+    private boolean isCallToOf(ExpressionTree expression) {
+        return OptionalUtils.isMethodInvocation(expression, "of", 1, atypeFactory);
+    }
+
+    /** @return true iff expression is a call to Optional creation: of, ofNullable. */
+    private boolean isOptionalCreation(MethodInvocationTree methInvok) {
+        return OptionalUtils.isMethodInvocation(methInvok, "of", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "ofNullable", 1, atypeFactory);
+    }
+
+    /**
+     * @return true iff expression is a call to Optional elimination: get, orElse, orElseGet,
+     *     orElseThrow.
+     */
+    private boolean isOptionalElimation(MethodInvocationTree methInvok) {
+        return OptionalUtils.isMethodInvocation(methInvok, "get", 0, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElse", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElseGet", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElseThrow", 1, atypeFactory);
+    }
+
+    @Override
+    public Void visitConditionalExpression(ConditionalExpressionTree node, Void p) {
+        handleTernaryIsPresentGet(node);
+        return super.visitConditionalExpression(node, p);
+    }
+
+    /**
+     * Part of rule #3.
+     *
+     * <p>Pattern match for: {@code VAR.isPresent() ? VAR.get().METHOD() : VALUE}
+     *
+     * <p>Prefer: {@code VAR.map(METHOD).orElse(VALUE);}
+     */
+    public void handleTernaryIsPresentGet(ConditionalExpressionTree node) {
+
+        ExpressionTree condExpr = TreeUtils.skipParens(node.getCondition());
+        ExpressionTree trueExpr = TreeUtils.skipParens(node.getTrueExpression());
+        ExpressionTree falseExpr = TreeUtils.skipParens(node.getFalseExpression());
+
+        if (!isCallToIsPresent(condExpr)) {
+            return;
+        }
+        ExpressionTree receiver = TreeUtils.getReceiverTree(condExpr);
+
+        if (trueExpr.getKind() != Kind.METHOD_INVOCATION) {
+            return;
+        }
+        ExpressionTree trueReceiver = TreeUtils.getReceiverTree(trueExpr);
+        if (!isCallToGet(trueReceiver)) {
+            return;
+        }
+        ExpressionTree getReceiver = TreeUtils.getReceiverTree(trueReceiver);
+
+        // What is a better way to do this than string comparison?
+        if (receiver.toString().equals(getReceiver.toString())) {
```

Comment:
This is OK for now. When going to transfer functions this would be solved using the correct Store entries. (Comparing the strings of the expressions also doesn't take care of side-effecting methods as receivers...)

Reply:
OK, we'll leave it for now and improve it in the future.  The comment was there so that during code review someone would tell me the right way to do it, because I knew this was not it. :-)